HeterogeneousCore/CUDACore/interface/CUDAScopedContext.h

```diff
@@ -0,0 +1,252 @@
+                                    edm::WaitingTaskWithArenaHolder waitingTaskHolder,
+                                    CUDAContextState& state)
+      : CUDAScopedContextGetterBase(streamID), holderHelper_{std::move(waitingTaskHolder)}, contextState_{&state} {}
+
+  /// Constructor to (possibly) re-use a CUDA stream (no need for context beyond acquire())
+  explicit CUDAScopedContextAcquire(const CUDAProductBase& data, edm::WaitingTaskWithArenaHolder waitingTaskHolder)
+      : CUDAScopedContextGetterBase(data), holderHelper_{std::move(waitingTaskHolder)} {}
+
+  /// Constructor to (possibly) re-use a CUDA stream, and the context is needed after acquire()
+  explicit CUDAScopedContextAcquire(const CUDAProductBase& data,
+                                    edm::WaitingTaskWithArenaHolder waitingTaskHolder,
+                                    CUDAContextState& state)
+      : CUDAScopedContextGetterBase(data), holderHelper_{std::move(waitingTaskHolder)}, contextState_{&state} {}
+
+  ~CUDAScopedContextAcquire();
+
+  template <typename F>
+  void pushNextTask(F&& f) {
+    if (contextState_ == nullptr)
+      throwNoState();
+    holderHelper_.pushNextTask(std::forward<F>(f), contextState_);
+  }
+
+  void replaceWaitingTaskHolder(edm::WaitingTaskWithArenaHolder waitingTaskHolder) {
+    holderHelper_.replaceWaitingTaskHolder(std::move(waitingTaskHolder));
+  }
+
+private:
+  void throwNoState();
+
+  impl::CUDAScopedContextHolderHelper holderHelper_;
+  CUDAContextState* contextState_ = nullptr;
+};
+
+/**
+ * The aim of this class is to do necessary per-event "initialization" in ExternalWork produce() or normal produce():
+ * - setting the current device
+ * - synchronizing between CUDA streams if necessary
+ * and enforce that those get done in a proper way in RAII fashion.
+ */
+class CUDAScopedContextProduce : public impl::CUDAScopedContextGetterBase {
+public:
+  /// Constructor to create a new CUDA stream (non-ExternalWork module)
+  explicit CUDAScopedContextProduce(edm::StreamID streamID) : CUDAScopedContextGetterBase(streamID) {}
+
+  /// Constructor to (possibly) re-use a CUDA stream (non-ExternalWork module)
+  explicit CUDAScopedContextProduce(const CUDAProductBase& data) : CUDAScopedContextGetterBase(data) {}
+
+  /// Constructor to re-use the CUDA stream of acquire() (ExternalWork module)
+  explicit CUDAScopedContextProduce(CUDAContextState& state)
+      : CUDAScopedContextGetterBase(state.device(), state.releaseStreamPtr()) {}
+
+  ~CUDAScopedContextProduce();
+
+  template <typename T>
+  std::unique_ptr<CUDAProduct<T>> wrap(T data) {
+    // make_unique doesn't work because of private constructor
+    //
+    // CUDAProduct<T> constructor records CUDA event to the CUDA
+    // stream. The event will become "occurred" after all work queued
+    // to the stream before this point has been finished.
+    std::unique_ptr<CUDAProduct<T>> ret(new CUDAProduct<T>(device(), streamPtr(), std::move(data)));
```

Comment:
`std::make_unique`?

Reply:
Doesn't work because the non-default constructors of `CUDAProduct<T>` are private (whether that is really necessary is then another topic).